sim/src/image.rs

```diff
@@ -104,22 +110,28 @@ impl ImagesBuilder {
     /// Construct an `Images` that doesn't expect an upgrade to happen.
     pub fn make_no_upgrade_image(self) -> Images {
         let mut flash = self.flash;
-        let primaries = install_image(&mut flash, &self.slots, 0, 32784, false);
-        let upgrades = install_image(&mut flash, &self.slots, 1, 41928, false);
+        let images = self.slots.into_iter().map(|slots| {
+            let primaries = install_image(&mut flash, &slots, 0, 32784, false);
+            let upgrades = install_image(&mut flash, &slots, 1, 41928, false);
+            OneImage {
+                slots: slots,
+                primaries: primaries,
+                upgrades: upgrades,
+            }}).collect();
         Images {
             flash: flash,
             areadesc: self.areadesc,
-            slots: self.slots,
-            primaries: primaries,
-            upgrades: upgrades,
+            images: images,
             total_count: None,
         }
     }
 
     /// Construct an `Images` for normal testing.
     pub fn make_image(self) -> Images {
         let mut images = self.make_no_upgrade_image();
-        mark_upgrade(&mut images.flash, &images.slots[1]);
+        for image in &images.images {
+            mark_upgrade(&mut images.flash, &image.slots[1]);
```

Comment:
This is OK, but semantics needs to be defined later, do we have to mark every upgrade slot?

Reply:
I'd like to write some additional tests that are able to upgrade one slot without the other. There also need to be tests that verify that dependencies are checked properly. But, currently, it always upgrades both slots.